CV2-4696 extract multiple media from tipline request

melsawy opened the pull request:
## Description

The request include on of the following types:

1. Long text (text with number of words > min_number_of_words_for_tipline_submit_shortcut) 
2. Short text (text with number of words <= min_number_of_words_for_tipline_submit_shortcut)
3. Media (image, audio, video, etc)

Based on the these types the created items should be

1. Item per each media (type 3)
2. Item per each long text (type 1)

The request text for created items include all text (long text, short text and media urls)

References: CV2-4696

## How has this been tested?

Implemented automated tests.

## Things to pay attention to during code review

Please describe parts of the change that require extra attention during code review, for example:

- File FFFF, line LL: This refactoring does this and this. Is it consistent with how it’s implemented elsewhere?
- Etc.

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have added unit and feature tests, if the PR implements a new feature or otherwise would benefit from additional testing
- [ ] I have added regression tests, if the PR fixes a bug
- [ ] I have added logging, exception reporting, and custom tracing with any additional information required for debugging
- [ ] I considered secure coding practices when writing this code. Any security concerns are noted above.
- [ ] I have commented my code in hard-to-understand areas, if any
- [ ] I have made needed changes to the README
- [ ] My changes generate no new warnings
- [ ] If I added a third party module, I included a rationale for doing so and followed our current [guidelines](https://meedan.atlassian.net/wiki/spaces/ENG/overview#Choose-the-%E2%80%9Cright%E2%80%9D-3rd-party-module)

caiosba reviewed (changes requested):
Sawy, I added a few comments but I would like to clarify one thing which was not clear for me from the code:

For this case, where a conversation bundle has this 5 messages:

1. Long text
2. Short text
3. Image
4. Video
5. Long text

Is it going to create 4 items with 4 tipline requests, for 1, 3, 4 and 5?

And in that case, the tipline request body will be the same for all of them, with this format?

```
Long text
Short text
https://image.url
https://video.url
Long text
```

melsawy commented on app/models/concerns/smooch_messages.rb:
```diff
@@ -278,15 +278,62 @@ def bundle_list_of_messages(list, last, reject_payload = false)
+      else
+        # Get text that match min_number_of_words_for_tipline_submit_shortcut(long text)
+        long_text = message_text.values
+        messages.each do |message|
```
Done

melsawy commented on app/models/concerns/smooch_messages.rb:
```diff
@@ -278,15 +278,62 @@ def bundle_list_of_messages(list, last, reject_payload = false)
-      if ['default_requests', 'irrelevant_search_result_requests'].include?(type)
-        self.process_message(bundle, app_id, send_message, type)
+    def bundle_list_of_messages_to_items(list, last, reject_payload = false)
+      # Sperate list into multiple messages based on media files and long text
```
Done

melsawy commented:
> Sawy, I added a few comments but I would like to clarify one thing which was not clear for me from the code:
> 
> For this case, where a conversation bundle has this 5 messages:
> 
> 1. Long text
> 2. Short text
> 3. Image
> 4. Video
> 5. Long text
> 
> Is it going to create 4 items with 4 tipline requests, for 1, 3, 4 and 5?
> 
> And in that case, the tipline request body will be the same for all of them, with this format?
> 
> ```
> Long text
> Short text
> https://image.url
> https://video.url
> Long text
> ```

yes @caiosba should create 4 items with same request body for all items and already I added a unit test to cover this case

melsawy commented on app/models/concerns/smooch_messages.rb:
```diff
@@ -278,15 +278,62 @@ def bundle_list_of_messages(list, last, reject_payload = false)
-      bundle = self.bundle_list_of_messages(list, last)
-      if ['default_requests', 'irrelevant_search_result_requests'].include?(type)
-        self.process_message(bundle, app_id, send_message, type)
+    def bundle_list_of_messages_to_items(list, last, reject_payload = false)
```
Removed `reject_payload`  argument

melsawy commented on app/models/concerns/smooch_messages.rb:
```diff
@@ -278,15 +278,62 @@ def bundle_list_of_messages(list, last, reject_payload = false)
+        if ['timeout_requests', 'menu_options_requests', 'resource_requests', 'relevant_search_result_requests', 'timeout_search_requests'].include?(type)
+          key = "smooch:banned:#{message['authorId']}"
+          if Rails.cache.read(key).nil?
+            [annotated].flatten.uniq.each_with_index { |a, i| self.save_message_later(message, app_id, type, a, i * 60) }
```
I depend on existing tests to cover this case but lmk how can I test this one using UI?

caiosba replied:
Hum... try to trigger a search that returns more than one result and vote that the results are relevant?

computermacgyver commented on app/models/concerns/smooch_messages.rb:
```diff
@@ -278,15 +278,60 @@ def bundle_list_of_messages(list, last, reject_payload = false)
+      else
+        # Attach all existing text (media text, long text and short text) to each item
+        # and add a boundary so we can easily split messages if needed
+        all_text = messages.collect{ |m| m['text'] }.concat(short_text).reject{ |t| t.blank? }.join("\n#{Bot::Smooch::MESSAGE_BOUNDARY}")
```
Does this put the messages in the correct order? For example, if we have:
Long text
Short text
Media 1

I think this code sets `all_text` to 
Long text
Media 1
Short text

But we want `all_text` to be
Long text
Short text
Media 1

caiosba replied:
@melsawy please review.

melsawy replied:
yes @computermacgyver my code set short text at the end, I'll set it in right order

melsawy replied:
Done

caiosba commented on app/models/concerns/smooch_messages.rb:
```diff
@@ -364,6 +409,8 @@ def save_message(message_json, app_id, author = nil, request_type = 'default_req
     end
 
     def smooch_save_tipline_request(message, associated, app_id, author, request_type, associated_obj)
+      message['text'] = message['request_body'] unless message['request_body'].blank?
+      message.delete('request_body')
```
Why we need this, Sawy?

melsawy replied:
I define a new key ` request_body`    to set all text the should attach to all items as by using the same `text` key I found two issues :
* Attach `media_url` to long text the item parsed as web-link due to existing URL and lead us to invalid item
* having multiple long text leads us to have multiple items created with same text.

caiosba replied:
Ah OK, thanks Sawy!

caiosba reviewed (approved):
Thanks Sawy for addressing the comments. There is a question from Scott and another question from me too.

caiosba reviewed (changes requested):
@melsawy , as per this ticket description, we agreed to not change how searching currently works, because of other product implications:

> We’re not going to change the search logic: we’re still going to search by the first media 
> only. The change here is only on how the request data is saved, not on how search results 
> are retrieved and returned to users.

Any other reason why you had to do this change or can you revert that?

melsawy commented:
@caiosba I reverted last commit
